keystore/v2/keystore/signature/notary.go

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package signature implements generation and verification of signatures
+// used by KeyStore to authenticate stored key data.
+package signature
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"hash"
+	"time"
+
+	"github.com/cossacklabs/acra/keystore/v2/keystore/asn1"
+	log "github.com/sirupsen/logrus"
+)
+
+const serviceName = "keystore"
+const notarySubsystemName = "notary"
+
+// Errors produced by signature verification:
+var (
+	ErrNoSignature    = fmt.Errorf("KeyStore: missing signature")
+	ErrSignatureError = fmt.Errorf("KeyStore: invalid signature")
+)
+
+// KeyedHMAC provides prekeyed HMAC algorithms used by Notary.
+type KeyedHMAC interface {
+	HmacSha256() hash.Hash
+}
+
+// Notary cryptographically signs provided ASN.1 data.
+type Notary struct {
+	hmac KeyedHMAC
+	log  *log.Entry
+}
+
+// NewNotary makes a new notary with given encryptor.
+func NewNotary(hmac KeyedHMAC) (*Notary, error) {
```

Comment:
why not to accept just hash.Hash?

Reply:
> why not to accept just `hash.Hash`?

Mostly for extensibility so that we can add a new hash functions without changing signatures and making them confusing. It is possible to have

```go
func NewNotary(sha256 hash.Hash)
```

but then you might accidentally pass a `sha256.New()` there instead of `hmac.New(sha256.New, yourKey)`. When the time comes to add a new function, this changes into

```go
func NewNotary(sha256, keccak hash.Hash)
```

which makes it easy to accidentally swap the functions. Even worse if we just replace it:

```go
func NewNotary(keccak hash.Hash)
```

the signature is the same, old code still compiles, but behaves incorrectly. In every case you need to go and manually hunt down every call to `NewNotary()` in the code base and make sure it’s correct.

_Actually_ it should not be a problem in practice because most likely there would be exactly one call to `NewNotary()` from the only key store we use. But I’d still be consistent with `KeyEncryptor`’s idea to keep _all_ cryptography behind a interface. Though, in case of HMAC-SHA-256 it’s kind of a leaky abstraction...

Okay! I can’t say you instantly convinced me, but you got me thinking. Replacing this with just `hash.Hash` keyed by the caller should probably allow us to simply reuse the existing `KeyEncryptor` interface and avoid inventing new ones. I’m still somewhat worried about accidentally using the wrong function. And about having to write `hmac.New()` manually instead of having some `NewSCellEncryptorFromEnvironment()` to sort everything out for you. But I’ll give it a thought.

Reply:
actually I dont like that we accept interface with hardcoded name of algorithm: hmac.
I think will be better if we will name interface like "Signer" or "SignAlgorithm" (or any other which will be abstract) and which will force manually implement some wrapper and disallow to use standard insecure hash functions. for example:
```
type Signer interface {
  hash.Hash
  Signer()
}
type KeyedHmac struct {
  hash.Hash
}
func NewKeyedHmac(key []byte)(*KeyedHmac, error){
  return KeyedHmac{hmac.New(sha256.New(), key)), nil
}
// Signer implement interface Signer
func (h *KeyedHmac) Signer(){}
```

Here we accept our own interface, disallow to pass something like sha256.New and force to manually write some wrapper and implement `Signer()` method to pass as object

Reply:
Hm... I like this approach. I wonder if it’s possible to make it even more simple.

As I asked myself a question “why not?” I remembered that I could not think up a suitable generic interface. The issue is that `Signature` object stores _the algorithm_ too as ASN.1 Object Identifier, along with signature value. I was thinking about something even more generic:

```go
type Signer interface {
        Sign(data, context []byte) []Signature
        Verify(data, context []byte, signatures []Signature) error
}
```

But in this case we will need to 1) store the algorithm in `Signature`:

```go
type Signature struct {
        Algorithm SignatureAlgorithm
        Value []byte
}

type SignatureAlgorithm int
const (
    SignatureSha256 SignatureAlgorithm = iota
)
```

2) map SignatureSha256 onto ASN.1 OID, 3) map ASN.1 OIDs back into `SignatureAlgorithm ` and construct a temporary `[]Signature`. I though it was too much of a hassle if the interface is still explicit about using SHA-256 (and thus does not really hide this detail).

P.S. I named it “notary” because I constantly confuse “singer” with “signer”.

Reply:
@Lagovas, after more and more thinking, it seems I've found a signature interface that should fill the bill. Please give a look to recently pushed changes.

We put cryptography behind the interface for two main reasons:

- make it possible to easily swap algorithms if necessary
- hide direct access to encryption/signature keys

Original interfaces succeeded in the second goal but not so much in the first one. That code duplicated existing Secure Cell encryptor code and leaked the implementation detail of using SHA-256. In the end, in order to add a new signature algorithm we would have to update a whole lot of interfaces, etc. it's hardly _easily_ extensible.

New interface consists of `signature.Algorithm` which defines abstract signing algorithm. It is used by `signature.Notary` to sign ASN.1 objects and verify signatures on them. You can pass a list of signatures for the notary to use. Now we can easily add new algorithms by implementing `signature.Algorithm` and adding it to the list.

With this we no longer need to duplicate `KeyEncryptor` interface and Secure Cell encryption. We can reuse existing implementations. We also reexport existing interface for convenience.

`crypto.SignSha256` provides default implementation of signature algorithm: HMAC-SHA-256. As a bonus, this interface allows us to avoid keeping the signature key in memory (we store only keyed `hash.Hash`).

We also keep convenience utilities: `crypto.KeyStoreSuite` (which is now a struct) to avoid passing tons of arguments to key store constuctors as well as `NewSCellSuite()` function which constructs the default cryptography suite: encrypt with Secure Cell, sign with HMAC-SHA-256.